README.md
# gmsmith [![Build status](https://travis-ci.org/twolfson/gmsmith.png?branch=master)](https://travis-ci.org/twolfson/gmsmith)

[GM][gm] engine for [spritesmith][].

[gm]: http://aheckmann.github.io/gm/
[spritesmith]: https://github.com/Ensighten/spritesmith

## Requirements
`gmsmith` depends on [gm][] which depends on [GraphicsMagick][].

I have found it is best to install from the site rather than through a package manager (e.g. `apt-get`) to get the latest as well as without transparency issues.

This module has been developed and tested against `1.3.17`.

> Alertnatively, you can use [ImageMagick][] which is implicitly discovered if `gm` is not installed.
> http://www.imagemagick.org/script/index.php

[ImageMagick]: http://www.imagemagick.org/script/index.php
[GraphicsMagick]: http://www.graphicsmagick.org/

## Getting Started
Install the module with: `npm install gmsmith`

```js
// Load in our dependencies
var Gmsmith = require('gmsmith');

// Create a new engine
var gmsmith = new Gmsmith();

// Interpret some images from disk
gmsmith.createImages(['img1.jpg', 'img2.png'], function handleImages (err, imgs) {
  // If there was an error, throw it
  if (err) {
    throw err;
  }

  // We recieve images in the same order they were given
  imgs[0].width; // 50 (pixels)
  imgs[0].height; // 100 (pixels)

  // Create a canvas that fits our images (200px wide, 300px tall)
  var canvas = gmsmith.createCanvas(200, 300);

  // Add the images to our canvas (at x=0, y=0 and x=50, y=100 respectively)
  canvas.addImage(imgs[0], 0, 0);
  canvas.addImage(imgs[1], 50, 100);

  // Export canvas to image
  var resultStream = canvas['export']({format: 'png'});
  resultStream; // Readable stream outputting PNG image of the canvas
});
```

## Documentation
This module was built to the specification for spritesmith engines.

**Specification version:** 2.0.0

https://github.com/twolfson/spritesmith-engine-spec/tree/2.0.0

### `new Engine(options)`
This is also known as `new Gmsmith`.

Our `Engine` constructor provides support for the following options:

- options `Object`
    - imagemagick `Boolean` - Flag to indicate whether to use [ImageMagick][] over [GraphicsMagick][]
        - When `true`, [ImageMagick][] will be used. Otherwise, [implicit discovery](#requirements) will be used.

### `engine.createImages(images, cb)`
Our `createImages` methods supports the following types of images:

- image `String` - Filepath to image
- image `Object` - Vinyl object with `null` for image (uses provided filepath)

We accept but don't fully support the following types of images:

- image `Object` - Vinyl object with buffer for image (ignores contents, uses provided filepath)
- image `Object` - Vinyl object with stream for image (ignores contents, uses provided filepath)

### `canvas.export(options)`
Our `export` method provides support for the following options:

- options `Object`
    - format `String` - Output image format to callback with. Currently, png and jpeg are available.
    - quality `Number` - Quality of output image on a scale from 0 to 100

## Contributing
In lieu of a formal styleguide, take care to maintain the existing coding style. Add unit tests for any new or changed functionality. Lint using `npm run lint` and test via `npm test`.

## Donating
Support this project and [others by twolfson][twolfson-projects] via [donations][twolfson-support-me].

<http://twolfson.com/support-me>

[twolfson-projects]: http://twolfson.com/projects
[twolfson-support-me]: http://twolfson.com/support-me

## License
Copyright (c) 2013 Todd Wolfson

Licensed under the MIT license.


CHANGELOG.md
# gmsmith changelog
1.4.0 - Upgraded dependencies via @striezel in #19

1.3.0 - Upgraded dependencies via @striezel in #17

1.2.0 - Upgraded to `gm@1.23.1` to fix vulnerabilities via @danez in #15

1.1.6 - Moved to Node.js>=4 to fix Travis CI

1.1.5 - Corrected CHANGELOG

1.1.4 - Upgraded to `spritesmith-engine-test@5.0.0` to verify Vinyl@2 support

1.1.3 - Replaced Gittip with support me page

1.1.2 - Added fallback for `options` via @@marekventur in #12

1.1.1 - Added `format` documentation to README via @marekventur in #11

1.1.0 - Updated dependencies via @danez in #9. Fixes #7

1.0.1 - Updated license in `package.json` to SPDX format via @danez in #8

1.0.0 - Upgraded to `spritesmith-engine-spec@2.0.0`

0.6.4 - Added `specVersion` to repository and `spritesmith-engine` to keywords

0.6.3 - Fixed failing Travis CI tests

0.6.2 - Cleaned up README and updated spec link to `spritesmith-engine-spec`

0.6.1 - Moved to `spritesmith-engine-test@3.0.0` to clean up technical debt

0.6.0 - Cleaned up technical debt (e.g. writing to disk, disjoint yet related parts of code)

0.5.4 - Removed `grunt` as a `devDependency`

0.5.3 - Added `twolfson-style` for linting

0.5.2 - Updated supported node versions to `>= 0.10.0`

0.5.1 - Added `foundry` for release

0.5.0 - Upgraded `gm` and `temporary` dependencies to add `node@0.12` and `iojs` support

0.4.5 - Added documentation for `quality`

0.4.4 - Upgraded to `spritesmith-engine-test@2.0.0` and moved to `mocha`

0.4.3 - Upgraded `npm` inside Travis CI to fix `node@0.8` issues

0.4.2 - Repaired legacy logic that caused 0.4.0 patch to fail whenever `.set` was called

0.4.1 - Re-enabling Travis CI tests for all supported node versions

0.4.0 - Implicitly discover `imagemagick` when `gm` doesn't exist and `imagemagick` flag is unspecified

0.3.0 - Added imagemagick to test suite and corrected `preinstall` for `package.json`

0.2.3 - Removed postinstall piping to allow cross-platform

0.2.2 - Integrated Travis CI

0.2.1 - Upgraded `doubleshot` and skipping `imagemagick` tests for approachability

0.2.0 - Added `get/set` support and `imagemagick` flag setting

0.1.3 - See `git log`


test/install.sh
#!/usr/bin/env bash
# Exit early on error
set -e

# Perform our installations
if test -n "$TEST_IMAGEMAGICK"; then
  # Install imagemagick
  sudo apt-get update
  sudo apt-get install imagemagick -y
else
  # Uninstall imagemagick from Travis CI
  sudo apt-get remove imagemagick -y

  # Download, make, and install graphicsmagick
  wget http://sourceforge.net/projects/graphicsmagick/files/graphicsmagick/1.3.36/GraphicsMagick-1.3.36.tar.xz/download --output-document GraphicsMagick-1.3.36.tar.xz
  tar xvf GraphicsMagick-1.3.36.tar.xz
  cd GraphicsMagick-1.3.36
  ./configure
  make
  sudo make install
  cd ..
fi


test/gmsmith_test.js
// Load our dependencies
var gmsmith = require('../');
var spritesmithEngineTest = require('spritesmith-engine-test');

// Configure gmsmith for our environment
// DEV: In case it recurs, we had downcasting of imagemagick's spritesheets to 8 bit for `get-pixels` loading
//   See https://github.com/twolfson/gmsmith/blob/0.4.3/test/gmsmith_test_content.js#L39
var engineOptions = {};
if (process.env.TEST_IMAGEMAGICK === 'TRUE') {
  engineOptions.imagemagick = true;
}

// Run our tests
spritesmithEngineTest.run({
  engine: gmsmith,
  engineName: 'gmsmith',
  engineOptions: engineOptions,
  tests: {
    // DEV: PNG seems to darken with only 1 image on IMAGEMAGICK, disable those tests for now
    renderPngCanvas: !(process.env.TEST_IMAGEMAGICK),
    // Disable buffer/stream tests since they are warnings only and we don't use contents
    renderPngBufferVinylCanvas: false,
    renderPngStreamVinylCanvas: false,
    // DEV: PNG seems to darken with only 1 image on IMAGEMAGICK, disable those tests for now
    renderPngNullVinylCanvas: !(process.env.TEST_IMAGEMAGICK),
    renderMultiplePngImages: true,
    // DEV: JPG seems to change significantly on Travis CI for IMAGEMAGICK, disable those tests for now
    renderJpgCanvas: !(process.env.TRAVIS && process.env.TEST_IMAGEMAGICK),
    renderGifCanvas: false,
    // If we are on Windows, skip over performance test (it cannot handle the long argument string)
    renderManyPngImages: process.platform === 'win32'
  }
});


test/gmsmith_test_content.js
// Load in parts to make our content
var exec = require('child_process').exec,
    async = require('async'),
    expect = require('chai').expect,
    deepEqual = require('deep-equal'),
    getPixels = require('get-pixels'),
    extend = require('obj-extend'),
    Tempfile = require('temporary/lib/file'),
    testContent = require('spritesmith-engine-test').content,
    smith = require('../lib/gmsmith');

// Duck punch over test items
var content = extend({}, testContent, {
  'gmsmith': function () {
    this.smith = smith;
    smith.clearSettings();
    if (process.env.TEST_IMAGEMAGICK === 'TRUE') {
      smith.set({imagemagick: true});
    } else if (process.env.TEST_IMAGEMAGICK === 'IMPLICIT_WITH_SET') {
      smith.set({});
    }

    var expectedDir = __dirname + '/expected_files';
    this.expectedFilepaths = [
      expectedDir + '/multiple-gm.png',
      expectedDir + '/multiple-im.png'
    ];
  },
  'can output an image': [function saveResultToFile (done) {
    this.tmpFile1 = new Tempfile();
    this.tmpFile1.path += '.png';
    this.tmpFile1.writeFile(this.result, 'binary', done);
  }, function adjustBitDepth (done) {
    this.actualFile = this.tmpFile1;
    if (process.env.TEST_IMAGEMAGICK) {
      this.tmpFile2 = new Tempfile();
      this.tmpFile2.path += '.png';
      this.actualFile = this.tmpFile2;
      exec('convert ' + this.tmpFile1.path + ' -depth 8 ' + this.tmpFile2.path, done);
    } else {
      done();
    }
  }, function getActualPixels (done) {
    var that = this;
    getPixels(this.actualFile.path, function (err, actualPixels) {
      that.actualPixels = actualPixels;
      done(err);
    });
  }, function assertExpectedImages (done) {
    // Assert the actual image is the same expected
    var actualPixels = this.actualPixels,
        matchesAnImage = false;

    // ANTI-PATTERN: Looping over set without identifiable lines for stack traces
    async.forEachSeries(this.expectedFilepaths, function testAgainstExpected (filepath, cb) {
      if (matchesAnImage) {
        return cb();
      }

      getPixels(filepath, function (err, expectedPixels) {
        if (err) {
          return cb(err);
        }

        matchesAnImage = deepEqual(actualPixels, expectedPixels);
        cb();
      });
    }, function (err) {
      if (err) { return done(err); }

      expect(matchesAnImage).to.equal(true);
      done();
    });
  }, function cleanupFiles () {
    this.tmpFile1.unlinkSync();

    if (this.tmpFile2) {
      this.tmpFile2.unlinkSync();
    }
  }]
});

// If we are on Windows, skip performance test items
if (process.platform === 'win32') {
  delete content["interpretting a ridiculous amount of images"];
  delete content["does not crash"];
  delete content["returns an image"];
}

// Export the content
module.exports = content;

lib/canvas.js
// Load in our dependencies
var assert = require('assert');

// Define our canvas constructor
function Canvas(width, height, gmsmith) {
  // Ignore width/height
  // DEV: Our canvas dimensions are dynamically defined by x/y of each image
  var canvas = gmsmith.gm(1, 1, 'transparent');

  // Override the -size options (won't work otherwise)
  canvas._in = ['-background', 'transparent'];

  // Save the canvas
  this.canvas = canvas;
}
Canvas.defaultFormat = 'png';
Canvas.supportedFormats = ['png', 'jpg', 'jpeg'];
Canvas.prototype = {
  addImage: function addImage (img, x, y) {
    // Add the image
    var canvas = this.canvas;

    // TODO: Pull request this in to gm
    canvas.out('-page');
    canvas.out('+' + x + '+' + y);
    canvas.out(img.file);
  },
  'export': function exportFn (options) {
    // Flatten the image (with transparency)
    var canvas = this.canvas;
    canvas.mosaic();

    // Validate our format
    var format = options.format || Canvas.defaultFormat;
    assert(Canvas.supportedFormats.indexOf(format) !== -1,
      '`gmsmith` doesn\'t support exporting a "' + format + '". Please export either a `png` or `jpeg`');

    // Update the quality of the canvas (if specified)
    var quality = options.quality;
    if (quality !== undefined) {
      canvas.quality(quality);
    }

    var progressive = options.progressive || false;
    if (progressive && (format === 'jpeg' || format === 'jpg')) {
      canvas.interlace('Line');
    }

    // Export our data via a stream
    // https://github.com/aheckmann/gm/blob/1.21.1/lib/command.js#L116-L144
    return canvas.stream(format);
  }
};

// Export our Canvas
module.exports = Canvas;


lib/engine.js
// Load in our dependencies
var async = require('async');
var which = require('which');
var _gm = require('gm');
var Canvas = require('./canvas');

// Determine if GraphicsMagick exists on our `PATH`
var gmExists = false;
try {
  gmExists = !!which.sync('gm');
} catch (e) {
  // Ignore errors for `gm` not existing
}

// Define our engine constructor
function Gmsmith(options) {
  options = options || {};
  this.gm = _gm;
  var useImageMagick = options.hasOwnProperty('imagemagick') ? options.imagemagick : !gmExists;
  if (useImageMagick) {
    this.gm = _gm.subClass({imageMagick: true});
  }
}
Gmsmith.specVersion = '2.0.0';
Gmsmith.prototype = {
  createCanvas: function (width, height) {
    // Generate our canvas and then callback with it in a bit
    var canvas = new Canvas(width, height, this);
    return canvas;
  },
  createImage: function (file, callback) {
    // If the file is an object (i.e. a vinyl object)
    if (typeof file === 'object') {
      // If it's not null, warn the users about loading excess content
      if (!file.isNull()) {
        console.warn('`gmsmith` doesn\'t support in-memory content ' +
          'but we saw in-memory content for "' + file.path + '". ' +
          'This can be disabled via `read: false` in `gulp.src`');
      }

      // Use the file path from the vinyl object
      file = file.path;
    }

    // Create the image
    var img = this.gm(file);

    // In series...
    async.waterfall([
      // Grab the size
      function getImgSize (cb) {
        img.size(cb);
      },
      function saveImgSize (size, cb) {
        // Create a structure for preserving the height and width of the image
        var imgFile = {
          height: size.height,
          width: size.width,
          file: file
        };

        // Callback with the imgFile
        cb(null, imgFile);
      }
    ], callback);
  },
  createImages: function (files, callback) {
    // Map the files into their image counterparts
    // DEV: Magic number of 10 to prevent file descriptor overuse
    // This does not affect perf -- 12 seconds with 300, 11.5 with 10 for 2000 images (derp)
    async.mapLimit(files, 10, this.createImage.bind(this), callback);
  }
};

// Export the engine
module.exports = Gmsmith;


lib/gmsmith.js
// Load in our dependencies
var async = require('async');
var assert = require('assert');
var concat = require('concat-stream');
var which = require('which');
var _gm = require('gm');
var engine = {};
var settings = {};

// Determine if GraphicsMagick exists on our `PATH`
var gmExists = false;
try {
  gmExists = !!which.sync('gm');
} catch (e) {
  // Ignore errors for `gm` not existing
}

// Helper function to set settings
// ANTI-PATTERN: We are treaing gmsmith as a singleton and should re-design to use `engine` as a constructor
function set(options) {
  // Save settings
  Object.getOwnPropertyNames(options).forEach(function saveOption (key) {
    settings[key] = options[key];
  });
}

// Helper method to clear all settings
function clear() {
  settings = {};
}

// Getter method for settings
function get(key) {
  return settings[key];
}

// Helper method for grabbing gm instance (graphicsmagick vs imagemagick)
function getGm() {
  var useImagemagick = get('imagemagick');
  if (useImagemagick || (useImagemagick === undefined && !gmExists)) {
    return _gm.subClass({imageMagick: true});
  } else {
    return _gm;
  }
}

// Expose get/set to engine
engine.get = get;
engine.set = set;
engine.clearSettings = clear;

function Canvas() {
  var gm = getGm();
  var canvas = gm(1, 1, 'transparent');

  // Override the -size options (won't work otherwise)
  canvas._in = ['-background', 'transparent'];

  // Save the canvas
  this.canvas = canvas;
}
// Define mapping for extensions to gm counterpart
Canvas.formatToExtMap = {
  png: 'png',
  'image/png': 'png',
  jpg: 'jpeg',
  jpeg: 'jpeg',
  'image/jpg': 'jpeg',
  'image/jpeg': 'jpeg'
};
Canvas.prototype = {
  addImage: function addImage (img, x, y, cb) {
    // Add the image
    var canvas = this.canvas;

    // TODO: Pull request this in to gm
    canvas.out('-page');
    canvas.out('+' + x + '+' + y);
    canvas.out(img.file);
  },
  'export': function exportFn (options, cb) {
    // Flatten the image (with transparency)
    var canvas = this.canvas;
    canvas.mosaic();

    // Render the item
    return this._exportCanvas(options, cb);
  },

  /**
   * gm exporter
   * @param {Object} options Options to export with
   * @param {Number} [options.quality] Quality of the exported item
   * @param {Function} cb Error-first callback to return binary image string to
   */
  _exportCanvas: function (options, cb) {
    // Extract our extension
    var format = options.format || 'png';
    var ext = Canvas.formatToExtMap[format];
    assert(ext, '`gmsmith` doesn\'t support exporting a "' + format + '". Please export either a `png` or `jpeg`');

    // Update the quality of the canvas (if specified)
    var canvas = this.canvas;
    var quality = options.quality;
    if (quality !== undefined) {
      canvas.quality(quality);
    }

    // Export our data via a stream
    // https://github.com/aheckmann/gm/blob/1.21.1/lib/command.js#L116-L144
    var retStream = canvas.stream(ext);
    retStream.on('error', cb);
    retStream.pipe(concat(function handleBuff (buff) {
      cb(null, buff.toString('binary'));
    }));
  }
};

// Expose Canvas to engine
engine.Canvas = Canvas;

// Define our async canvas creator
// DEV: Our canvas is dynamically defined by x/y of each image
function createCanvas(width, height, callback) {
  // Generate our canvas and then callback with it in a bit
  var canvas = new Canvas();
  process.nextTick(function handleNextTick () {
    callback(null, canvas);
  });
}
engine.createCanvas = createCanvas;

// Write out Image as a static property of Canvas
/**
 * @param {String} file File path to load in
 * @param {Function} callback Error first callback to retrun the image from
 * @prop {Number} image.width
 * @prop {Number} image.height
 * @note Must be guaranteed to integrate into own library via .addImage
 */
function createImage(file, callback) {
  // Create the image
  var gm = getGm();
  var img = gm(file);

  // In series...
  async.waterfall([
    // Grab the size
    function getImgSize (cb) {
      img.size(cb);
    },
    function saveImgSize (size, cb) {
      // Create a structure for preserving the height and width of the image
      var imgFile = {
        height: size.height,
        width: size.width,
        file: file
      };

      // Callback with the imgFile
      cb(null, imgFile);
    }
  ], callback);
}
engine.createImage = createImage;

function createImages(files, callback) {
  // Map the files into their image counterparts
  // DEV: Magic number of 10 to prevent file descriptor overuse
  // This does not affect perf -- 12 seconds with 300, 11.5 with 10 for 2000 images (derp)
  async.mapLimit(files, 10, createImage, callback);
}
engine.createImages = createImages;

// Export the canvas
engine.specVersion = '1.1.0';
module.exports = engine;
